Allow attributes on record fields

mateuszbujalski opened the pull request:
This change extends the record syntax to allow decorating fields with attributes. 

I still need to write a test where I query the attribute from a tactic to make sure it works correctly.

mateuszbujalski commented on src/parser/FStar.Parser.ToDocument.fs:
```diff
@@ -973,7 +973,7 @@ and p_aqual = function
     str "#[" ^^ p_term false false t ^^ str "]" ^^ break1
   | Meta (Arg_qualifier_meta_attr t) ->
-    str "#[@" ^^ p_term false false t ^^ str "]" ^^ break1
+    str "#[@@" ^^ p_term false false t ^^ str "]" ^^ break1
```
I think the "#[@" syntax is deprecated? Is that correct?

nikswamy replied:
Agreed, good to fix this. thanks.

mateuszbujalski commented on src/parser/parse.fsy:
```diff
@@ -223,49 +224,49 @@ option___anonymous_1_:
+option___anonymous_10_:
+  
+    {    ( None )}
+| LBRACE noSeqTerm RBRACE
```
This doesn't seem related to my change. Not sure why this was generated.

mateuszbujalski commented on src/syntax/FStar.Syntax.Resugar.fs:
```diff
@@ -1190,7 +1190,17 @@ let resugar_typ env datacon_ses se : sigelts * A.tycon =
+                        let q = 
+                            // TODO: There's a 'FIXME' note on resugar_arg_qual that I don't fully understand. It doesn't seem applicable in this case as the 
+                            //       syntax for records only support 'Arg_qualifier_meta_attr' and the only problematic case seems to be 'inaccessible implicit argument of a data constructor'
+                            match resugar_arg_qual env qual with
```
@nikswamy This is my main question. The only supported qual for record field is an attribute, and looking at how 'resugar_arg_qual' works, it should always return 'Some'. Do you think this code is fine? Should I at least leave some comment here?

mateuszbujalski commented:
@nikswamy I've extended the test and querying seems to work fine. I consider this one done and ready for review :)

nikswamy commented on tests/micro-benchmarks/RecordFieldAttributes.fst:
```diff
@@ -0,0 +1,111 @@
+        #[@@ (description "This is a number") ] field1 : int;
+        #[@@ (description "This is a string") ] field2 : string;
+    }
+
```
Sadly, there's one aspect of it that I neglected to notice earlier. These attributes on binders are unfortunately coupled with implicit arguments (i.e., the preceding `#`) makes both these fields implicits arguments to the record constructor.

That means that the code below fails to typecheck

```
@@expect_failure]
let test (x:int) (y:string) = { field1 = x; field2 = y }
```

And you instead have to write:

```
let test (x:int) (y:string) : r = Mkr #x #y
```

I think we should decouple attributes on binders with implicit arguments. i.e., you should be able to associate an attribute with a binder whether it is implicit or not. E.g., 

```
([@@ attr1] x : t) -> (#[@@attr2] y:t) -> ... 
```

The first being an explicit and the second an implicit arg.

@aseemr @R1kM  ... wdyt? This should be fine for Steel too, right?

R1kM replied:
I think it should be fine for Steel indeed. So far, I don't think we needed the framing attribute on explicit arguments, but this would be good to have for the future

mateuszbujalski replied:
Right, I completely missed this.

"I think we should decouple attributes on binders with implicit arguments." 
@nikswamy Do I understand correctly that 'Meta' aqual cases are currently assumed to be implicit? And I should extend the aqual with some new case which could be used for an explicit argument? Plus change the rule so that when HASH is missing, we go with the new case?

nikswamy replied:
Yes, that's right.

mateuszbujalski commented:
@nikswamy @aseemr I've updated the PR after the refactoring done by Aseem. I wrote some simple test for records and now it works both for implicit and explicit fields. I'll add some more tests for various cases where attributes are now possible. 

Unfortunately, I wasn't able to change the parser without increasing the number of shift/reduce conflicts around LBRACK_AT_AT (the total number went up from 6 to 11 - I've temporarily hacked the makefile to accept this to see if CI will be happy with the rest of the changes). What's currently pushed there seems to work but it's probably best if it's reverted and reworked. It would be great if you could take a look :)
